saq/queue.py

```diff
@@ -393,12 +412,122 @@ async def enqueue(self, job_or_func, **kwargs):
         job.queued = now()
         job.status = Status.QUEUED
 
-        if not await self._enqueue_script(
+        on_enqueue_generators = [
+            on_enqueue(job) for on_enqueue in self._on_enqueue.values()
+        ]
+        for gen in on_enqueue_generators:
+            await run_iterator_to_first_yield(gen)
+
+        enqueued = await self._enqueue_script(
             keys=[self._incomplete, job.id, self._queued, job.abort_id],
             args=[self.serialize(job), job.scheduled],
             client=self.redis,
-        ):
-            return None
+        )
+
+        for gen in on_enqueue_generators:
+            await run_iterator_to_completion(gen)
+
+        if enqueued:
+            logger.info("Enqueuing %s", job)
+            return job
+        return None
+
+    async def apply(self, job_or_func, timeout=None, **kwargs):
+        """
+        Enqueue a job and wait for its result.
+
+        If the job is successful, this returns its result.
+        If the job is unsuccessful, this raises a JobError.
+
+        Example::
+            try:
+                assert await queue.apply("add", a=1, b=2) == 3
+            except JobError:
+                print("job failed")
+
+        job_or_func: Same as Queue.enqueue
+        timeout: Seconds to wait for job to complete. If None or 0, wait forever.
+        kwargs: Same as Queue.enqueue
+        """
+        job = await self.enqueue(job_or_func, **kwargs)
+        return await job.wait_for_result(timeout)
+
+    async def map(self, job_or_func, iter_kwargs, timeout=None, **kwargs):
+        """
+        Enqueue multiple jobs and collect all of their results.
+
+        This raises a JobError if any of the jobs are unsuccessful.
+
+        This waits for job results in the order they are enqueued, so JobExceptions are
```

Comment:
I was running into issues when running large amounts of jobs (10k).

I figured it had to do with large amounts of Redis subscriptions. I didn't spend too much time debugging because I figured waiting for jobs in order was less resource intensive and probably good enough in most cases.

I can dig more into this if you feel it's necessery.

Reply:
i see this can be solved with a semaphore

https://github.com/tobymao/saq/blob/baa37aa6938ccbc8f7e4fd81d6b9c4ffabdbf5e8/benchmarks/simple.py#L11

it doesn't make much of a difference for waiting for the results, but certainly enqueing as on line 487 is much more efficient with an async gather as opposed to for loop

if it's not too much work can we change it use a semaphore, that might also make the code cleaner because then you can just use a wait for instead of a custom time check

Reply:
can we make subscribe more efficient by reusing the same subscription ?

Reply:
Ah yeah I like that idea.
Let me rework this.

Reply:
re: the semaphore
I think this is something that aioredis can handle for us with it's `BlockingConnectionPool`.
Then we don't need to litter our code with semaphore acquisition.

Reply:
Although we might need some documentation for users on how/when to use such a pool.

Reply:
ah interesting, can we make that the default? i didn't realize they had one

Reply:
sorry if it's a bit  more work, but can you test the simple bench marks with blocking connection pool so there's no degradation in perf over a semaphore?